### PR TITLE
[spdx] Fix several bugs

### PR DESCRIPTION
**Pull request by Thomas1664**
Fixed bugs:

- Parsing `vcpkg_from_git` didn't work because it used the results of parsing `vcpkg_from_github`
- `vcpkg_from_bitbucket` and `vcpkg_from_gitlab` weren't parsed
- Indexing of `SPDXRef-resource` started at 1
- Only the first occurence of one of `vcpkg_from_*` / `vcpkg_download_distfile` was parsed
  I.e., if there were 2 occurences of `vcpkg_from_github`, only the first one was included in the SBOM file.

Changes in behavior:

- `find_cmake_invocation` no longer uses case insensitive comparison

### CODE REVIEW

File: src/vcpkg/cmake.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            return {it + 1, std::find(it + 1, command.end(), '"')};
+        }
+        return {it, std::find_if(it + 1, command.end(), [](char ch) {
+                    return ParserBase::is_whitespace(ch) || ch == ')';
+                })};
+    }
+    std::string replace_cmake_var(StringView text, StringView var, StringView value)
+    {
+        return Strings::replace_all(text, fmt::format("${{{}}}", var), value);
```
Comment (ras0219-msft) [inline src/vcpkg/cmake.cpp:36]:
```suggestion
        return Strings::replace_all(text, std::string("${") + var + "}", value);
```
I think `fmt::format` is more obfuscating than clarifying here.

File: include/vcpkg/cmake.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <vcpkg/base/stringview.h>
+
+namespace vcpkg
+{
+    StringView find_cmake_invocation(StringView content, StringView command);
```
Comment (ras0219-msft) [inline include/vcpkg/cmake.h:7]: Given that these functions are _very much_ heuristics and only usable with caveats, I don't think they should be introduced as an independent "cmake" module. Instead, I think they should remain in spdx.cpp and exposed as needed for testing purposes.

Ideally, we will get rid of these heuristics entirely and instead introduce some sort of "side channel" that allows the portfile to record accesses -- such as by writing out a json document inside `vcpkg_from_github()` with the actual information.

If we do introduce another area that needs CMake parsing heuristics before we're able to delete them from the SPDX path, we can move them out into separate files at that point.

File: src/vcpkg-test/cmake.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        auto res = find_cmake_invocation("lorem_ipsum(", "lorem_ipsum");
+        REQUIRE(res.empty());
+    }
+    {
+        auto res = find_cmake_invocation("lorem_ipum()", "lorem_ipsum");
+        REQUIRE(res.empty());
+    }
+    {
+        auto res = find_cmake_invocation("lorem_ipsum( )", "lorem_ipsum");
```
Comment (ras0219-msft) [inline src/vcpkg-test/cmake.cpp:39]: This should test additional edge cases:
1. `lorem_ipsum_`
2. `lorem_ipsum (`
3. `lorem_ipsum` (EOF)

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Overall, I am really happy with the improvements in this PR! However, I think there's a few aspects of the implementation that need to be addressed.
 
This PR adds additional SPDX features but no SPDX tests. I'd like to see tests covering `vcpkg::run_resource_heuristics` for each of the new helpers. Adding tests to cover the existing `vcpkg_from_github` would be appreciated but not necessary :)

Additionally, can you please include a sample of SPDX documents before and after your change for a handful of ports, demonstrating that all changes were intentional and positive?

> Indexing of SPDXRef-resource started at 1

This is not a bug, unless you can reference an SPDX specification that indicates it should start at 0?

**Review by BillyONeal — APPROVED** [review]

File: src/vcpkg/spdx.cpp [hunk 4]
```diff
@@ -34,36 +117,12 @@ static void append_move_if_exists_and_array(Json::Array& out, Json::Object& obj,
     }
 }
 
-static StringView find_cmake_invocation(StringView contents, StringView command)
-{
-    auto it = Strings::case_insensitive_ascii_search(contents, command);
-    if (it == contents.end()) return {};
-    it += command.size();
-    if (it == contents.end()) return {};
-    if (ParserBase::is_word_char(*it)) return {};
```
Comment (BillyONeal) [inline src/vcpkg/spdx.cpp:43]: @ras0219-msft I really don't understand the significance of these 'word char' checks here. If a file has `vcpkg_from_github .... vcpkg_from_git`, then we should skip over the first one and go to the real one, which I added a test for and fixed. Please double check that I did not misunderstand these' purpose.

**Comment by BillyONeal** [comment]
@Thomas1664 Thanks for the fixes!
